packages/keyboardfocus/src/useKeyboardFocus.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright Zendesk, Inc.
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { useState } from 'react';
+
+import { composeEventHandlers } from '@zendeskgarden/container-selection';
+
+export function useKeyboardFocus() {
+  const [keyboardFocused, setKeyboardFocused] = useState(false);
+  let keyboardFocusable = true;
+
+  const onPointerDown = () => {
+    keyboardFocusable = false;
+  };
+
+  const onMouseDown = () => {
+    onPointerDown();
+
+    /**
+     * This is necessary to recognize focus events caused by keyboard vs mouseDown.
+     * Due to React event ordering this is always called before onFocus.
+     */
+    setTimeout(() => {
+      keyboardFocusable = true;
+    }, 0);
+  };
+
+  const onFocus = () => {
+    if (keyboardFocusable) {
+      setKeyboardFocused(true);
+    }
+  };
+
+  const onBlur = () => {
+    setKeyboardFocused(false);
+  };
+
+  const getFocusProps = ({ tabIndex = 0, ...props } = {}) => {
+    return {
+      tabIndex,
+      onBlur: composeEventHandlers(props.onBlur, onBlur),
+      onFocus: composeEventHandlers(props.onFocus, onFocus),
+      onMouseDown: composeEventHandlers(props.onMouseDown, onMouseDown),
```

Comment:
To help solve some of our Menu -> Button focus issues, should we consider moving some of these `onMouseDown`, `onPointerDown`, and `onTouchStart` events to the `document/window` level?

The hooks could use an `environment` prop to allow the iframe usages within Web Widget if we need the flexibility.  [Downshift uses it to good effect](https://github.com/downshift-js/downshift#environment)

Reply:
Can you elaborate on the menu -> button focus issues?

So if we basically implement the w3c polyfill that should fix the issues?

Reply:
Are you referencing the [WICG implementation](https://github.com/WICG/focus-visible)? I didn't realize that's how they were handling it.

I'm not sure having _every_ handler on the document is necessary for our use-case, but we could move some of them to https://github.com/downshift-js/downshift/blob/master/src/downshift.js#L1089-L1093 to better handle clicks that occur outside the boundary of the component (like a Menu).

I was also thinking of how we can more easily control the ability to show/hide that focus when the programmatic `focus()` is used. For some implementations, it might be easier to use the [useEmperativeHandle()](https://reactjs.org/docs/hooks-reference.html#useimperativehandle) hook to customize the `focus()` method.

Reply:
I think we're confusing what this is trying to solve with other logic that is handled in menus. 

To me this hook is only concerned with the element that has `getFocusedProps` spread onto it if it's focused via keyboard or some other input device.

I'm not sure if attaching to the document is a good idea or I'm missing some nuance here that you're thinking about.

Reply:
No, you're right. This hook could be composed with another custom hook to achieve that functionality.

LGTM!